File: sem4/MPP/mpp-proiect-repository-mariaruncan/ProjJavaFX/Persistence/src/main/java/app/persistence/repository/jdbc/GamesDBRepository.java
package app.persistence.repository.jdbc;

import app.model.Game;
import app.model.Team;
import app.persistence.GamesRepository;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.sql.*;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

@Component
public class GamesDBRepository implements GamesRepository {

    private final JdbcUtils dbUtils;
    private static final Logger logger = LogManager.getLogger();

    public GamesDBRepository(Properties props){
        logger.info("Initializing GamesDBRepository with properties: {} ", props);
        dbUtils = new JdbcUtils(props);
    }

    public GamesDBRepository(){
        Properties props = new Properties();
        try {
            props.load(EmployeesDBRepository.class.getResourceAsStream("/bd.properties"));
            System.out.println("Server properties set. ");
            props.list(System.out);
        } catch (IOException e) {
            System.err.println("Cannot find server.properties " + e);
            dbUtils = null;
            return;
        }
        dbUtils = new JdbcUtils(props);
    }

    @Override
    public void add(Game elem) {
        logger.traceEntry("saving task {} ", elem);
        Connection con = dbUtils.getConnection();
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("d-MMM-yy");

        try(PreparedStatement preStm = con.prepareStatement("INSERT INTO Games (team1Id, team2Id, description, " +
                "totalNoOfTickets, soldTickets, pricePerTicket, date) VALUES (?, ?, ?, ?, ?, ?, ?);")){
            preStm.setInt(1, Math.toIntExact(elem.getTeam1().getId()));
            preStm.setInt(2, Math.toIntExact(elem.getTeam2().getId()));
            preStm.setString(3, elem.getDescription());
            preStm.setInt(4, elem.getTotalNoOfTickets());
            preStm.setInt(5, elem.getSoldTickets());
            preStm.setDouble(6, elem.getPricePerTicket());
            preStm.setString(7, elem.getDate().format(formatter));
            int result = preStm.executeUpdate();
            logger.trace("Saved {} instances", result);
        }
        catch(SQLException ex){
            logger.error(ex);
            System.err.println("Error DB" + ex);
        }
        logger.traceExit();
    }

    @Override
    public void delete(Long id) {
        logger.traceEntry("deleting task {} ", id);
        Connection con = dbUtils.getConnection();

        try(PreparedStatement preStm = con.prepareStatement("DELETE FROM Games WHERE id = ?;")){
            preStm.setInt(1, Math.toIntExact(id));
            int result = preStm.executeUpdate();
            logger.trace("Deleted {} instances", result);
        }
        catch(SQLException ex){
            logger.error(ex);
            System.err.println("Error DB" + ex);
        }
        logger.traceExit();
    }

    @Override
    public void update(Long aLong, Game elem) {
        logger.traceEntry("updating task {} ", elem);
        Connection con = dbUtils.getConnection();
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("d-MMM-yy");

        try(PreparedStatement preStm = con.prepareStatement("UPDATE Games SET team1Id = ?, team2Id = ?, " +
                "description = ?, totalNoOfTickets = ?, soldTickets = ?, pricePerTicket = ?, date = ?" +
                " WHERE id = ?;")){
            preStm.setInt(1, Math.toIntExact(elem.getTeam1().getId()));
            preStm.setInt(2, Math.toIntExact(elem.getTeam2().getId()));
            preStm.setString(3, elem.getDescription());
            preStm.setInt(4, elem.getTotalNoOfTickets());
            preStm.setInt(5, elem.getSoldTickets());
            preStm.setDouble(6, elem.getPricePerTicket());
            preStm.setString(7, elem.getDate().format(formatter));
            preStm.setLong(8, elem.getId());
            int result = preStm.executeUpdate();
            logger.trace("Updated {} instances", result);
        }
        catch(SQLException ex){
            logger.error(ex);
            System.err.println("Error DB" + ex);
        }
        logger.traceExit();
    }

    @Override
    public Game findById(Long aLong) {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        Game game = null;
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("d-MMM-yy");

        try(PreparedStatement preStm = con.prepareStatement("""
                SELECT * FROM Games G
                INNER JOIN Teams T1 on G.team1Id = T1.id
                INNER JOIN Teams T2 on G.team2Id = T2.id
                WHERE G.id = ?
                LIMIT 1;""")){
            preStm.setInt(1, Math.toIntExact(aLong));
            try(ResultSet result = preStm.executeQuery()){
                if(result.next()){
                    Long team1Id = result.getLong(9);
                    String team1Name = result.getString(10);
                    Team team1 = new Team(team1Name);
                    team1.setId(team1Id);

                    Long team2Id = result.getLong(11);
                    String team2Name = result.getString(12);
                    Team team2 = new Team(team2Name);
                    team2.setId(team2Id);

                    Long gameId = result.getLong(1);
                    String description = result.getString("description");
                    Integer totalNoOfTickets = result.getInt("totalNoOfTickets");
                    Integer soldTickets = result.getInt("soldTickets");
                    Float pricePerTicket = result.getFloat("pricePerTicket");
                    LocalDate date = LocalDate.parse(result.getString("date").split(" ")[0], formatter);

                    game = new Game(team1, team2, date, description, totalNoOfTickets, soldTickets, pricePerTicket);
                    game.setId(gameId);
                }
            }
        }
        catch(SQLException ex){
            logger.error(ex);
            System.err.println("Error DB" + ex);
        }
        logger.traceExit();
        return game;
    }

    @Override
    public Iterable<Game> findAll() {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        List<Game> games = new ArrayList<>();
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("d-MMM-yy");

        try(PreparedStatement preStm = con.prepareStatement("""
                SELECT * FROM Games G
                INNER JOIN Teams T1 on G.team1Id = T1.id
                INNER JOIN Teams T2 on G.team2Id = T2.id;""")){
            try(ResultSet result = preStm.executeQuery()){
                while(result.next()){
                    Long team1Id = result.getLong(9);
                    String team1Name = result.getString(10);
                    Team team1 = new Team(team1Name);
                    team1.setId(team1Id);

                    Long team2Id = result.getLong(11);
                    String team2Name = result.getString(12);
                    Team team2 = new Team(team2Name);
                    team2.setId(team2Id);

                    Long gameId = result.getLong(1);
                    String description = result.getString("description");
                    Integer totalNoOfTickets = result.getInt("totalNoOfTickets");
                    Integer soldTickets = result.getInt("soldTickets");
                    Float pricePerTicket = result.getFloat("pricePerTicket");
                    String dateStr = result.getString("date").split(" ")[0];
                    LocalDate date = LocalDate.parse(dateStr, formatter);

                    Game game = new Game(team1, team2, date, description, totalNoOfTickets, soldTickets, pricePerTicket);
                    game.setId(gameId);
                    games.add(game);
                }
            }
        }
        catch(SQLException ex){
            logger.error(ex);
            System.err.println("Error DB" + ex);
        }
        logger.traceExit();
        return games;
    }
}

File: sem4/MPP/mpp-proiect-repository-mariaruncan/ProjCSharp/Persistence/db/GamesDBRepository.cs
using log4net;
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SQLite;
using Model;

namespace Persistence.db
{
    public class GamesDBRepository : GamesRepository
    {
        private static readonly ILog logger = LogManager.GetLogger(typeof(EmployeesDBRepository));
        DBUtils dbUtils;

        public GamesDBRepository()
        {
            logger.Info("Initializing GamesDBRepository");
            dbUtils = new DBUtils();
        }

        public void add(Game elem)
        {
            logger.Info($"saving game {elem}");
            IDbConnection con = dbUtils.GetConnection();

            try
            {
                using (IDbCommand cmd = con.CreateCommand())
                {
                    cmd.CommandText = "INSERT INTO Games (team1Id, team2Id, description, totalNoOfTickets, soldTickets, pricePerTicket, date) VALUES (@team1Id, @team2Id, @description, @totalNoOfTickets, @soldTickets, @pricePerTicket, @date);";

                    IDbDataParameter paramTeam1Id = cmd.CreateParameter();
                    paramTeam1Id.ParameterName = "@team1Id";
                    paramTeam1Id.Value = elem.Team1.Id;
                    cmd.Parameters.Add(paramTeam1Id);

                    IDbDataParameter paramTeam2Id = cmd.CreateParameter();
                    paramTeam2Id.ParameterName = "@team2Id";
                    paramTeam2Id.Value = elem.Team2.Id;
                    cmd.Parameters.Add(paramTeam2Id);

                    IDbDataParameter paramDescription = cmd.CreateParameter();
                    paramDescription.ParameterName = "@description";
                    paramDescription.Value = elem.Description;
                    cmd.Parameters.Add(paramDescription);

                    IDbDataParameter paramTotalNoOfTickets = cmd.CreateParameter();
                    paramTotalNoOfTickets.ParameterName = "@totalNoOfTickets";
                    paramTotalNoOfTickets.Value = elem.TotalNoOfTickets;
                    cmd.Parameters.Add(paramTotalNoOfTickets);

                    IDbDataParameter paramSoldTickets = cmd.CreateParameter();
                    paramSoldTickets.ParameterName = "@soldTickets";
                    paramSoldTickets.Value = elem.SoldTickets;
                    cmd.Parameters.Add(paramSoldTickets);

                    IDbDataParameter paramPricePerTicket = cmd.CreateParameter();
                    paramPricePerTicket.ParameterName = "@pricePerTicket";
                    paramPricePerTicket.Value = elem.PricePerTicket;
                    cmd.Parameters.Add(paramPricePerTicket);
                    
                    IDbDataParameter paramDate = cmd.CreateParameter();
                    paramDate.ParameterName = "@date";
                    paramDate.Value = elem.Date.ToString();
                    cmd.Parameters.Add(paramDate);

                    var result = cmd.ExecuteNonQuery();
                    if (result == 0)
                        throw new RepositoryException("No game added!");
                    logger.Info($"saved {result} instances");
                }
            }
            catch (SQLiteException ex)
            {
                logger.Error(ex.Message);
                Console.WriteLine("Error DB " + ex.Message);
            }
        }

        public void delete(long id)
        {
            logger.Info($"deleting game {id}");
            IDbConnection con = dbUtils.GetConnection();

            try
            {
                using (IDbCommand cmd = con.CreateCommand())
                {
                    cmd.CommandText = "DELETE FROM Games WHERE id = @id;";
                    IDbDataParameter paramId = cmd.CreateParameter();
                    paramId.ParameterName = "@id";
                    paramId.Value = id;
                    cmd.Parameters.Add(paramId);

                    var dataR = cmd.ExecuteNonQuery();
                    if (dataR == 0)
                        throw new RepositoryException("No game deleted!");
                    logger.Info("exit delete EmployeeRepo");
                }
            }
            catch (SQLiteException ex)
            {
                logger.Error(ex.Message);
                Console.WriteLine("Error DB " + ex.Message);
            }
        }

        public IEnumerable<Game> findAll()
        {
            logger.Info("enter findAll");
            IDbConnection con = dbUtils.GetConnection();
            IList<Game> games = new List<Game>();

            try
            {
                using (IDbCommand cmd = con.CreateCommand())
                {
                    cmd.CommandText = "SELECT G.id, G.description, G.totalNoOfTickets, G.soldTickets, " +
                        "G.pricePerTicket, G.date, T1.id, T1.name, T2.id, T2.name " +
                        "FROM Games G " +
                        "INNER JOIN Teams T1 on G.team1Id = T1.id " +
                        "INNER JOIN Teams T2 on G.team2Id = T2.id; ";
                    
                    using (var result = cmd.ExecuteReader())
                    {
                        while (result.Read())
                        {
                            long id = result.GetInt64(0);
                            String description = result.GetString(1);
                            int totalNoOfTickets = result.GetInt32(2);
                            int soldTickets = result.GetInt32(3);
                            float pricePerTicket = result.GetFloat(4);
                            DateTime date = DateTime.Parse(result.GetString(5));
                            long team1Id = result.GetInt64(6);
                            String team1Name = result.GetString(7);
                            long team2Id = result.GetInt64(8);
                            String team2Name = result.GetString(9);
                            Team team1 = new Team(team1Name);
                            team1.Id = team1Id;
                            Team team2 = new Team(team2Name);
                            team2.Id = team2Id;
                            Game g = new Game(team1, team2, date, description, totalNoOfTickets, soldTickets, pricePerTicket);
                            g.Id = id;
                            games.Add(g);
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
            logger.Info("exit findAll");
            return games;
        }

        public Game findById(long id)
        {
            logger.Info("enter findById");
            IDbConnection con = dbUtils.GetConnection();
            Game game = null;

            try
            {
                using (IDbCommand cmd = con.CreateCommand())
                {
                    cmd.CommandText = "SELECT G.id, G.description, G.totalNoOfTickets, G.soldTickets, " +
                            "G.pricePerTicket, G.date, T1.id, T1.name, T2.id, T2.name " +
                            "FROM Games G " +
                            "INNER JOIN Teams T1 on G.team1Id = T1.id " +
                            "INNER JOIN Teams T2 on G.team2Id = T2.id " +
                            "WHERE G.id = @id;";
                    IDbDataParameter paramId = cmd.CreateParameter();
                    paramId.ParameterName = "@id";
                    paramId.Value = id;
                    cmd.Parameters.Add(paramId);

                    using (var result = cmd.ExecuteReader())
                    {
                        if (result.Read())
                        {
                            long idG = result.GetInt64(0);
                            String description = result.GetString(1);
                            int totalNoOfTickets = result.GetInt32(2);
                            int soldTickets = result.GetInt32(3);
                            float pricePerTicket = result.GetFloat(4);
                            DateTime date = DateTime.Parse(result.GetString(5));
                            long team1Id = result.GetInt64(6);
                            String team1Name = result.GetString(7);
                            long team2Id = result.GetInt64(8);
                            String team2Name = result.GetString(9);
                            Team team1 = new Team(team1Name);
                            team1.Id = team1Id;
                            Team team2 = new Team(team2Name);
                            team2.Id = team2Id;
                            game = new Game(team1, team2, date, description, totalNoOfTickets, soldTickets, pricePerTicket);
                            game.Id = id;
                        }
                    }
                    logger.InfoFormat("exit findById");
                    return game;
                }
            }
            catch(SQLiteException ex)
            {
                logger.Error(ex.Message);
                Console.WriteLine("Error DB " + ex.Message);
                return game;
            }
        }

        public void update(Game elem, long id)
        {
            logger.Info($"updating game {elem}");
            IDbConnection con = dbUtils.GetConnection();

            try
            {
                using (IDbCommand cmd = con.CreateCommand())
                {
                    cmd.CommandText = "UPDATE Games SET team1Id = @team1Id, team2Id = @team2Id, " +
                        "description = @description, totalNoOfTickets = @totalNoOfTickets, " +
                        "soldTickets = @soldTickets, pricePerTicket = @pricePerTicket , date = @date " +
                        "WHERE id = @id;";
                    IDbDataParameter paramTeam1Id = cmd.CreateParameter();
                    paramTeam1Id.ParameterName = "@team1Id";
                    paramTeam1Id.Value = elem.Team1.Id;
                    cmd.Parameters.Add(paramTeam1Id);

                    IDbDataParameter paramTeam2Id = cmd.CreateParameter();
                    paramTeam2Id.ParameterName = "@team2Id";
                    paramTeam2Id.Value = elem.Team2.Id;
                    cmd.Parameters.Add(paramTeam2Id);

                    IDbDataParameter paramDescription = cmd.CreateParameter();
                    paramDescription.ParameterName = "@description";
                    paramDescription.Value = elem.Description;
                    cmd.Parameters.Add(paramDescription);

                    IDbDataParameter paramTotalNoOfTickets = cmd.CreateParameter();
                    paramTotalNoOfTickets.ParameterName = "@totalNoOfTickets";
                    paramTotalNoOfTickets.Value = elem.TotalNoOfTickets;
                    cmd.Parameters.Add(paramTotalNoOfTickets);

                    IDbDataParameter paramsoldTickets = cmd.CreateParameter();
                    paramsoldTickets.ParameterName = "@soldTickets";
                    paramsoldTickets.Value = elem.SoldTickets;
                    cmd.Parameters.Add(paramsoldTickets);

                    IDbDataParameter paramPricePerTicket = cmd.CreateParameter();
                    paramPricePerTicket.ParameterName = "@pricePerTicket";
                    paramPricePerTicket.Value = elem.PricePerTicket;
                    cmd.Parameters.Add(paramPricePerTicket);
                    
                    IDbDataParameter paramDate = cmd.CreateParameter();
                    paramDate.ParameterName = "@date";
                    paramDate.Value = elem.Date.ToString();
                    cmd.Parameters.Add(paramDate);

                    IDbDataParameter paramId = cmd.CreateParameter();
                    paramId.ParameterName = "@id";
                    paramId.Value = elem.Id;
                    cmd.Parameters.Add(paramId);

                    var result = cmd.ExecuteNonQuery();
                    logger.Info($"updated {result} instances");
                }
            }
            catch(SQLiteException ex)
            {
                logger.Error(ex.Message);
                Console.WriteLine("Error DB " + ex.Message);
            }
        }
    }
}
